// @dhtmlx/trial-react-gantt/src/wx/Counter.js
import React, { Component } from "react";
import cls from "classnames";
import css from "./Counter.module.css";

export default class Counter extends Component {
  constructor(props) {
    super();
    this.step = props.step || 1;
    this.min = props.min || 1;
    this.max = Infinity;
    this.digits = new RegExp("^[0-9]+$");
    this.error = false;
  }

  inc() {
    if (this.props.value >= this.max) return;
    const v = this.props.value + this.step;
    if (this.props.change) this.props.change(v);
  }

  dec() {
    if (this.props.value <= this.min) return;
    const v = this.props.value - this.step;
    if (this.props.change) this.props.change(v);
  }

  change(e) {
    let v = e.target.value;
    if (this.digits.test(v)) {
      this.error = false;
      if (this.props.change) this.props.change(v);
    } else {
      this.error = true;
    }
  }

  render() {
    const { label, value } = this.props;

    return (
      <div className={css.counter}>
        <label className={css.label}>{label}</label>
        <div className={css.controls}>
          <button
            className={cls(css.btn, css.btnDec)}
            onClick={() => this.dec()}
          >
            <svg
              className={css.dec}
              width="12"
              height="2"
              viewBox="0 0 12 2"
              fill="none"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path d="M11.2501 1.74994H0.750092V0.249939H11.2501V1.74994Z" />
            </svg>
          </button>

          <input
            className={cls(css.input, css.error ? this.error : "")}
            type="text"
            value={value}
            onChange={(e) => this.change(e)}
          />

          <button
            className={cls(css.btn, css.btnInc)}
            onClick={() => this.inc()}
          >
            <svg
              className={css.inc}
              width="12"
              height="12"
              viewBox="0 0 12 12"
              fill="none"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.2501
					6.74994H6.75009V11.2499H5.25009V6.74994H0.750092V5.24994H5.25009V0.749939H6.75009V5.24994H11.2501V6.74994Z"
              />
            </svg>
          </button>
        </div>
      </div>
    );
  }
}
